examples/puzzles/tictactoe/tictactoe.qnt

```diff
@@ -38,104 +25,138 @@ module tictactoe {
     Set((3,1), (2,2), (1,3))
   )
 
-  def Won(player) = winningPatterns.exists(pattern =>
-    pattern.forall(coordinate => BoardIs(coordinate, player))
-  )
-
-  action MoveO = all {
-    nextTurn == "O",
-    not(Won("X")),
-    MoveToEmpty("O"),
-    nextTurn' = "X",
-  }
-
   val corners = Set(
     (1,1),
     (3,1),
     (1,3),
     (3,3)
   )
 
-  val BoardEmpty = tuples(1.to(3), 1.to(3)).forall(position =>
-    BoardIs(position, "_")
+  def hasValue(coordinate, player) =
+    board.get(coordinate._1).get(coordinate._2) == player
+
+  def isEmpty(coordinate) = hasValue(coordinate, "_")
+
+  val boardEmpty = tuples(1.to(3), 1.to(3)).forall(isEmpty)
+
+  def won(player) = winningPatterns.exists(pattern =>
+    pattern.forall(coordinate => hasValue(coordinate, player))
   )
 
-  action StartInCorner =
-    nondet corner = oneOf(corners)
-    Move("X", corner)
+  val boardFull = not(tuples(1.to(3), 1.to(3)).exists(isEmpty))
+  val stalemate = boardFull and not(won("X")) and not(won("O"))
+  val gameOver = won("X") or won("O") or stalemate
 
-  def canWinWithPattern(winningPattern) = and {
-    winningPattern.filter(p => BoardIs(p, "X")).size() == 2,
-    winningPattern.filter(p => BoardIs(p, "_")).size() == 1,
+  def canWinWithPattern(pattern) = and {
+    pattern.filter(coordinate => coordinate.hasValue("X")).size() == 2,
+    pattern.filter(coordinate => coordinate.isEmpty()).size() == 1,
   }
 
-  def canBlockWithPattern(winningPattern) = and {
-    winningPattern.filter(p => BoardIs(p, "O")).size() == 2,
-    winningPattern.filter(p => BoardIs(p, "_")).size() == 1,
+  def canBlockWithPattern(pattern) = and {
+    pattern.filter(coordinate => coordinate.hasValue("O")).size() == 2,
```

Comment:
I've considered this but opted for not doing it for two reasons:
1. I personally think it's easy to read `"X"` and `"O"`
2. The `player` word is overloaded, since the opponent is also a player. This is a problem also because I'm already using the name `player` as parameters for some definitions, so I'd have to find another word to refer to any of the players and I couldn't think of anything good enough.